PhysicsTools/PatUtils/interface/SmearedJetProducerT.h

```diff
@@ -21,370 +17,259 @@
+            const JetCollection& jets = *jets_collection;
+
+            if (m_genJetMatcher)
+                m_genJetMatcher->getTokens(event);
+
+            std::unique_ptr<JetCollection> smearedJets(new JetCollection());
+
+            for (const auto& jet: jets) {
+
+                if (! m_enabled) {
+                    // Module disabled. Simply copy the input jet.
+                    T smearedJet = jet;
+                    smearedJets->push_back(smearedJet);
+
+                    continue;
+                }
+
+                double jet_resolution = resolution.getResolution({{JME::Binning::JetPt, jet.pt()}, {JME::Binning::JetEta, jet.eta()}, {JME::Binning::Rho, *rho}});
+                double jer_sf = resolution_sf.getScaleFactor({{JME::Binning::JetEta, jet.eta()}}, m_systematic_variation);
+
+                if (m_debug) {
+                    std::cout << "jet:  pt: " << jet.pt() << "  eta: " << jet.eta() << "  phi: " << jet.phi() << "  e: " << jet.energy() << std::endl;
+                    std::cout << "resolution: " << jet_resolution << std::endl;
+                    std::cout << "resolution scale factor: " << jer_sf << std::endl;
+                }
+
+                const reco::GenJet* genJet = nullptr;
+                if (m_genJetMatcher)
+                    genJet = m_genJetMatcher->match(jet, jet.pt() * jet_resolution);
+
+                double smearFactor = 1.;
+
+                if (genJet) {
+                    /*
+                     * Case 1: we have a "good" gen jet matched to the reco jet
+                     */
+
+                    if (m_debug) {
+                        std::cout << "gen jet:  pt: " << genJet->pt() << "  eta: " << genJet->eta() << "  phi: " << genJet->phi() << "  e: " << genJet->energy() << std::endl;
+                    }
+
+                    double dE = jet.energy() - genJet->energy();
+                    smearFactor = 1 + (jer_sf - 1.) * dE / jet.energy();
+
+                } else if (jer_sf > 1) {
+                    /*
+                     * Case 2: we don't have a gen jet. Smear jet energy using a random gaussian variation
+                     */
+
+                    double sigma = jet_resolution * std::sqrt(jer_sf * jer_sf - 1);
+                    if (m_debug) {
+                        std::cout << "gaussian width: " << sigma << std::endl;
+                    }
+
+                    std::normal_distribution<> d(0, sigma);
+                    smearFactor = 1. + d(m_random_generator) / jet.energy();
+                } else if (m_debug) {
+                    std::cout << "Impossible to smear this jet" << std::endl;
+                }
+
+                T smearedJet = jet;
+                smearedJet.scaleEnergy(smearFactor);
+
+                if (m_debug) {
+                    std::cout << "smeared jet (" << smearFactor << "):  pt: " << smearedJet.pt() << "  eta: " << smearedJet.eta() << "  phi: " << smearedJet.phi() << "  e: " << smearedJet.energy() << std::endl;
```

Comment:
all "cout" use better be converted to LogPrint {on by default and doesn't add extra formatting}.
If this ever needs to be debugged in multithreaded mode, the "cout" parts will turn to junk in the log file.